test/vegeta/pacers/combined_pacer.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pacers
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	vegeta "github.com/tsenart/vegeta/lib"
+)
+
+// combinedPacer is a Pacer that combines multiple Pacers and runs them sequentially when being used for attack.
+type combinedPacer struct {
+	// pacers is a list of pacers that will be used sequentially for attack, must be more than 1 pacer.
+	pacers []vegeta.Pacer
+	// durations is the list of durations for the given Pacers, must have the same length as Pacers.
+	durations []time.Duration
+
+	// totalDuration is sum of the given Durations.
+	totalDuration uint64
+	// stepDurations is the accumulative duration of each step calculated from the given Durations.
+	stepDurations []uint64
+
+	curtPacerIndex  uint
+	prevElapsedHits uint64
+	prevElapsedTime uint64
+}
+
+// NewCombined returns a new CombinedPacer with the given config.
+func NewCombined(pacers []vegeta.Pacer, durations []time.Duration) (vegeta.Pacer, error) {
+	if len(pacers) == 0 || len(durations) == 0 || len(pacers) != len(durations) || len(pacers) == 1 {
+		return nil, errors.New("configuration for this CombinedPacer is invalid")
+	}
+
+	var totalDuration uint64
+	stepDurations := make([]uint64, len(pacers))
+	for i, duration := range durations {
+		totalDuration += uint64(duration)
+		if i == 0 {
+			stepDurations[i] = uint64(duration)
+		} else {
+			stepDurations[i] = stepDurations[i-1] + uint64(duration)
+		}
+	}
+	pacer := &combinedPacer{
+		pacers:    pacers,
+		durations: durations,
+
+		totalDuration: totalDuration,
+		stepDurations: stepDurations,
+
+		curtPacerIndex:  0,
```

Comment:
You don't really need to set the `0` values.

Reply:
Removed.

Reply:
I think they're still there.

Reply:
Sorry I missed it. Done.